### ISSUE TITLE
[BUG] When using a custom LLM as judge model (endpoints:/endpoint_name), response is NaN

### ISSUE THREAD

**Issue by ai-learner-00** [mlflow/mlflow#11087]
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Local machine

### Willingness to contribute

No. I cannot contribute a bug fix at this time.

### MLflow version

- Client: 2.10.0
- Tracking server: 2.10.0

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: WSL (Ubuntu)
- **Python version**: 3.10.13
- **yarn version, if running the dev UI**:

### Describe the problem

I am trying to use a [custom LLM as judge model](https://mlflow.org/docs/latest/llms/llm-evaluate/index.html#selecting-the-llm-as-judge-model), but using `endpoints:/chat` doesn't work. The resulting metrics are NaN.
![image](https://github.com/mlflow/mlflow/assets/158079614/6c6579d0-c30d-4e42-ab9f-3bfd3f0384e1)

By default, is the chat or completions version of gpt-4 being used?

`config.yaml`
```
endpoints:
  - name: completions
    endpoint_type: llm/v1/completions
    model:
      provider: openai
      name: gpt-3.5-turbo
      config:
        openai_api_key: $OPENAI_API_KEY
    limit:
      renewal_period: minute
      calls: 10

  - name: chat
    endpoint_type: llm/v1/chat
    model:
      provider: openai
      name: gpt-3.5-turbo
      config:
        openai_api_key: $OPENAI_API_KEY

  - name: azure-chat
    endpoint_type: llm/v1/chat
    model:
      provider: openai
      name: gpt-4
      config:
        openai_api_type: "azure"
        openai_api_key: $AZURE_OPENAI_API_KEY
        openai_deployment_name: "gpt-4"
        openai_api_base: "_"
        openai_api_version: "2023-05-15"

  - name: azure-completions
    endpoint_type: llm/v1/completions
    model:
      provider: openai
      name: gpt-4
      config:
        openai_api_type: "azure"
        openai_api_key: $AZURE_OPENAI_API_KEY
        openai_deployment_name: "gpt-4"
        openai_api_base: "_"
        openai_api_version: "2023-05-15"

  - name: embeddings
    endpoint_type: llm/v1/embeddings
    model:
      provider: openai
      name: text-embedding-ada-002
      config:
        openai_api_key: $OPENAI_API_KEY
```
Create deployments server
```bash
mlflow deployments start-server --config-path config.yaml --port 7000
```
Connect tracking server and deployments server
```bash
export MLFLOW_DEPLOYMENTS_TARGET="http://127.0.0.1:7000"
mlflow server --port 5000
```

I can use the endpoints via `get_deploy_client`, so I am not sure what went wrong.
```python
from mlflow.deployments import get_deploy_client

client = get_deploy_client("http://localhost:7000")
response = client.predict(
    endpoint="chat",
    inputs={"messages": [{"role": "user", "content": "Tell me a joke about rabbits"}]},
)
print(response)
```
```
{'id': 'chatcmpl-8qR3CU9RyUMeOPQZJsfr2WPbBaX5Q', 'object': 'chat.completion', 'created': 1707507730, 'model': 'gpt-3.5-turbo-0613', 'choices': [{'index': 0, 'message': {'role': 'assistant', 'content': 'Why did the rabbit go to the barber?\n\nBecause it wanted a hare-cut!'}, 'finish_reason': 'stop'}], 'usage': {'prompt_tokens': 13, 'completion_tokens': 16, 'total_tokens': 29}}
```
### Tracking information

```shell
System information: Windows 10.0.19045
Python version: 3.10.13
MLflow version: 2.10.0
MLflow module location: [c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\mlflow\__init__.py](file:///C:/Users/_/miniconda3/envs/ml-3.10/lib/site-packages/mlflow/__init__.py)
Tracking URI: http://127.0.0.1:5000/
Registry URI: http://127.0.0.1:5000/
MLflow dependencies: 
  Flask: 3.0.1
  Jinja2: 3.1.3
  aiohttp: 3.9.2
  alembic: 1.13.1
  boto3: 1.34.37
  botocore: 1.34.37
  click: 8.1.7
  cloudpickle: 3.0.0
  databricks-cli: 0.18.0
  docker: 7.0.0
  entrypoints: 0.4
  fastapi: 0.109.0
  gitpython: 3.1.41
  importlib-metadata: 6.11.0
  kubernetes: 29.0.0
  markdown: 3.5.2
  matplotlib: 3.8.2
  numpy: 1.26.3
  packaging: 23.2
  pandas: 2.2.0
  protobuf: 4.23.4
  pyarrow: 15.0.0
  pydantic: 2.6.0
  pytz: 2023.4
  pyyaml: 6.0.1
  querystring-parser: 1.2.4
  requests: 2.31.0
  scikit-learn: 1.4.0
  scipy: 1.12.0
  slowapi: 0.1.8
  sqlalchemy: 2.0.25
  sqlparse: 0.4.4
  tiktoken: 0.5.2
  uvicorn: 0.27.0.post1
  waitress: 2.1.2
  watchfiles: 0.21.0
```

### Code to reproduce issue

```python
from datasets import load_dataset

dataset = load_dataset("squad")
eval_df = dataset["train"].to_pandas().sample(n=10, random_state=0)
def get_answer(row):
    answers = row["answers"]["text"]
    return answers[0]

eval_df["answer"] = eval_df.apply(get_answer, axis=1)
eval_df = eval_df.drop(columns=["answers", "id", "title"])

import mlflow

mlflow.set_tracking_uri("http://127.0.0.1:5000")
mlflow.set_experiment("test-deployment")


from langchain_core.prompts import ChatPromptTemplate, PromptTemplate
from langchain_core.output_parsers import StrOutputParser
from langchain_core.runnables import RunnablePassthrough
from langchain_core.runnables import RunnableLambda, RunnablePassthrough
from langchain_openai import ChatOpenAI

template = """Answer the question based only on the following context:
{context}

Question: {question}
"""
prompt = ChatPromptTemplate.from_template(template)

llm = ChatOpenAI()
chain = (
    RunnablePassthrough()
    | prompt 
    | llm
    | StrOutputParser()
)
chain.invoke({"question": "What is 1 + 1?", "context": "1 + 1 is 3"})

from mlflow.metrics.genai import EvaluationExample, answer_similarity

# Create an example to describe what answer_similarity means like for this problem.
example = EvaluationExample(
    input="What is MLflow?",
    output="MLflow is an open-source platform for managing machine "
    "learning workflows, including experiment tracking, model packaging, "
    "versioning, and deployment, simplifying the ML lifecycle.",
    score=4,
    justification="The definition effectively explains what MLflow is "
    "its purpose, and its developer. It could be more concise for a 5-score.",
    grading_context={
        "targets": "MLflow is an open-source platform for managing "
        "the end-to-end machine learning (ML) lifecycle. It was developed by Databricks, "
        "a company that specializes in big data and machine learning solutions. MLflow is "
        "designed to address the challenges that data scientists and machine learning "
        "engineers face when developing, training, and deploying machine learning models."
    },
)

# Construct the metric using OpenAI GPT-4 as the judge
mlflow.deployments.set_deployments_target("http://localhost:7000")

judge_model = "endpoints:/chat"
# judge_model = "openai:/gpt-4"

answer_similarity_metric = answer_similarity(model=judge_model, examples=[example])

mlflow.deployments.get_deployments_target() # wasn't set even if in env variable

def model(input_df):
    answers = []
    for _, row in input_df.iterrows():
        answer = chain.invoke({"question": row["question"], "context": row["context"]})
        answers.append(answer)
    return answers

with mlflow.start_run(run_name="mistral"):
    mlflow.evaluate(
        model,
        eval_df,
        # model_type="question-answering",
        # evaluators="default",
        targets="answer",
        extra_metrics=[answer_similarity_metric], # faithfulness_metric, relevance_metric],
        evaluator_config={
            "col_mapping": {
                "inputs": "question",
            }
        },
    )
```

### Stack trace

```
[c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\mlflow\data\digest_utils.py:29](file:///C:/Users/_/miniconda3/envs/ml-3.10/lib/site-packages/mlflow/data/digest_utils.py:29): FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  string_columns = trimmed_df.columns[(df.applymap(type) == str).all(0)]
[c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\mlflow\models\evaluation\base.py:414](file:///C:/Users/_/miniconda3/envs/ml-3.10/lib/site-packages/mlflow/models/evaluation/base.py:414): FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  data = data.applymap(_hash_array_like_element_as_bytes)
[c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\mlflow\models\evaluation\base.py:414](file:///C:/Users/_/miniconda3/envs/ml-3.10/lib/site-packages/mlflow/models/evaluation/base.py:414): FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  data = data.applymap(_hash_array_like_element_as_bytes)
2024/02/09 15:08:14 INFO mlflow.models.evaluation.base: Evaluating the model with the default evaluator.
2024/02/09 15:08:14 INFO mlflow.models.evaluation.default_evaluator: Computing model predictions.
2024/02/09 15:08:23 INFO mlflow.models.evaluation.default_evaluator: Testing metrics on first row...
100%
1/1 [00:02<00:00, 2.03s/it]
[c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\numpy\core\fromnumeric.py:3504](file:///C:/Users/_/miniconda3/envs/ml-3.10/lib/site-packages/numpy/core/fromnumeric.py:3504): RuntimeWarning: Mean of empty slice.
  return _methods._mean(a, axis=axis, dtype=dtype,
[c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\numpy\core\_methods.py:129](file:///C:/Users/_/miniconda3/envs/ml-3.10/lib/site-packages/numpy/core/_methods.py:129): RuntimeWarning: invalid value encountered in scalar divide
  ret = ret.dtype.type(ret / rcount)
[c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\numpy\core\fromnumeric.py:3787](file:///C:/Users/_/miniconda3/envs/ml-3.10/lib/site-packages/numpy/core/fromnumeric.py:3787): RuntimeWarning: Degrees of freedom <= 0 for slice
  return _methods._var(a, axis=axis, dtype=dtype, out=out, ddof=ddof,
[c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\numpy\core\_methods.py:163](file:///C:/Users/_/miniconda3/envs/ml-3.10/lib/site-packages/numpy/core/_methods.py:163): RuntimeWarning: invalid value encountered in divide
  arrmean = um.true_divide(arrmean, div, out=arrmean,
[c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\numpy\core\_methods.py:198](file:///C:/Users/_/miniconda3/envs/ml-3.10/lib/site-packages/numpy/core/_methods.py:198): RuntimeWarning: invalid value encountered in scalar divide
  ret = ret.dtype.type(ret / rcount)
2024/02/09 15:08:25 INFO mlflow.models.evaluation.default_evaluator: Evaluating metrics: answer_similarity
100%
10/10 [00:00<00:00, 42.78it/s]
```

### Other info / logs

```
```

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [x] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [x] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by ai-learner-00** [comment]
When I only perform `mlflow ui` without starting a deployments server, there is no error. 

```python
with mlflow.start_run(run_name="mistral"):
    mlflow.evaluate(
        model,
        eval_df,
        # model_type="question-answering",
        # evaluators="default",
        targets="answer",
        extra_metrics=[answer_similarity_metric], # faithfulness_metric, relevance_metric],
        evaluator_config={
            "col_mapping": {
                "inputs": "question",
            }
        },
    )
```
```
c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\mlflow\data\digest_utils.py:29: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  string_columns = trimmed_df.columns[(df.applymap(type) == str).all(0)]
c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\mlflow\models\evaluation\base.py:414: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  data = data.applymap(_hash_array_like_element_as_bytes)
c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\mlflow\models\evaluation\base.py:414: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  data = data.applymap(_hash_array_like_element_as_bytes)
2024/02/12 14:46:15 INFO mlflow.models.evaluation.base: Evaluating the model with the default evaluator.
2024/02/12 14:46:15 INFO mlflow.models.evaluation.default_evaluator: Computing model predictions.
2024/02/12 14:46:37 INFO mlflow.models.evaluation.default_evaluator: Testing metrics on first row...
100%
1/1 [01:25<00:00, 85.78s/it]
c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\numpy\core\fromnumeric.py:3504: RuntimeWarning: Mean of empty slice.
  return _methods._mean(a, axis=axis, dtype=dtype,
c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\numpy\core\_methods.py:129: RuntimeWarning: invalid value encountered in scalar divide
  ret = ret.dtype.type(ret / rcount)
c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\numpy\core\fromnumeric.py:3787: RuntimeWarning: Degrees of freedom <= 0 for slice
  return _methods._var(a, axis=axis, dtype=dtype, out=out, ddof=ddof,
c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\numpy\core\_methods.py:163: RuntimeWarning: invalid value encountered in divide
  arrmean = um.true_divide(arrmean, div, out=arrmean,
c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\numpy\core\_methods.py:198: RuntimeWarning: invalid value encountered in scalar divide
  ret = ret.dtype.type(ret / rcount)
2024/02/12 14:48:03 INFO mlflow.models.evaluation.default_evaluator: Evaluating metrics: answer_similarity
100%
10/10 [01:27<00:00, 3.34s/it]
```

If I use the deploy client, I do get an error. (which is the expected behaviour)
```python
from mlflow.deployments import get_deploy_client

client = get_deploy_client("http://localhost:7000")
endpoints = client.list_endpoints()
for endpoint in endpoints:
    print(endpoint)
```
```
(Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x0000027B65441690>: Failed to establish a new connection: [WinError 10061] No connection could be made because the target machine actively refused it'))

During handling of the above exception, another exception occurred:

ConnectionError                           Traceback (most recent call last)
File c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\mlflow\utils\rest_utils.py:108, in http_request(host_creds, endpoint, method, max_retries, backoff_factor, backoff_jitter, extra_headers, retry_codes, timeout, raise_on_status, **kwargs)
    107 try:
--> 108     return _get_http_response_with_retries(
    109         method,
    110         url,
    111         max_retries,
    112         backoff_factor,
    113         backoff_jitter,
    114         retry_codes,
    115         raise_on_status,
    116         headers=headers,
    117         verify=host_creds.verify,
    118         timeout=timeout,
    119         **kwargs,
    120     )
    121 except requests.exceptions.Timeout as to:

File c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\mlflow\utils\request_utils.py:212, in _get_http_response_with_retries(method, url, max_retries, backoff_factor, backoff_jitter, retry_codes, raise_on_status, allow_redirects, **kwargs)
    210 allow_redirects = env_value if allow_redirects is None else allow_redirects
--> 212 return session.request(method, url, allow_redirects=allow_redirects, **kwargs)

File c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\requests\sessions.py:589, in Session.request(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)
    588 send_kwargs.update(settings)
--> 589 resp = self.send(prep, **send_kwargs)
    591 return resp

File c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\requests\sessions.py:703, in Session.send(self, request, **kwargs)
    702 # Send the request
--> 703 r = adapter.send(request, **kwargs)
    705 # Total elapsed time of the request (approximately)

File c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\requests\adapters.py:519, in HTTPAdapter.send(self, request, stream, timeout, verify, cert, proxies)
    517         raise SSLError(e, request=request)
--> 519     raise ConnectionError(e, request=request)
    521 except ClosedPoolError as e:

ConnectionError: HTTPConnectionPool(host='localhost', port=7000): Max retries exceeded with url: /api/2.0/endpoints/ (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x0000027B65441690>: Failed to establish a new connection: [WinError 10061] No connection could be made because the target machine actively refused it'))

During handling of the above exception, another exception occurred:

MlflowException                           Traceback (most recent call last)
Cell In[2], line 1
----> 1 endpoints = client.list_endpoints()
      2 for endpoint in endpoints:
      3     print(endpoint)

File c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\mlflow\deployments\mlflow\__init__.py:229, in MlflowDeploymentClient.list_endpoints(self)
    227 next_page_token = None
    228 while True:
--> 229     page = self._list_endpoints(next_page_token)
    230     endpoints.extend(page)
    231     next_page_token = page.token

File c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\mlflow\deployments\mlflow\__init__.py:181, in MlflowDeploymentClient._list_endpoints(self, page_token)
    179 def _list_endpoints(self, page_token=None) -> "PagedList[Endpoint]":
    180     params = None if page_token is None else {"page_token": page_token}
--> 181     response_json = self._call_endpoint(
    182         "GET", MLFLOW_DEPLOYMENTS_CRUD_ENDPOINT_BASE, json_body=params
    183     )
    184     routes = [
    185         Endpoint(
    186             **{
   (...)
    194         for resp in response_json.get("endpoints", [])
    195     ]
    196     next_page_token = response_json.get("next_page_token")

File c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\mlflow\deployments\mlflow\__init__.py:132, in MlflowDeploymentClient._call_endpoint(self, method, route, json_body, timeout)
    129 else:
    130     call_kwargs["json"] = json_body
--> 132 response = http_request(
    133     host_creds=_get_default_host_creds(self.target_uri),
    134     endpoint=route,
    135     method=method,
    136     timeout=MLFLOW_HTTP_REQUEST_TIMEOUT.get() if timeout is None else timeout,
    137     retry_codes=MLFLOW_DEPLOYMENT_CLIENT_REQUEST_RETRY_CODES,
    138     raise_on_status=False,
    139     **call_kwargs,
    140 )
    141 augmented_raise_for_status(response)
    142 return response.json()

File c:\Users\_\miniconda3\envs\ml-3.10\lib\site-packages\mlflow\utils\rest_utils.py:130, in http_request(host_creds, endpoint, method, max_retries, backoff_factor, backoff_jitter, extra_headers, retry_codes, timeout, raise_on_status, **kwargs)
    128     raise InvalidUrlException(f"Invalid url: {url}") from iu
    129 except Exception as e:
--> 130     raise MlflowException(f"API request to {url} failed with exception {e}")

MlflowException: API request to http://localhost:7000/api/2.0/endpoints/ failed with exception HTTPConnectionPool(host='localhost', port=7000): Max retries exceeded with url: /api/2.0/endpoints/ (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x0000027B65441690>: Failed to establish a new connection: [WinError 10061] No connection could be made because the target machine actively refused it'))
```

**Comment by harupy** [comment]
@ai-learner-00 This looks like a bug, investigating now.

**Comment by harupy** [comment]
@ai-learner-00 I filed #11096 to fix this bug.

**Comment by daanalfa** [comment]
> @ai-learner-00 I filed #11096 to fix this bug.

Adding this here as this is relevant to this issue by virtue of being a bug when using a 'custom LLM as judge model`.

@harupy if we use your fix in the PR #11096 with an Anthropic-Claude-v2 model via AWS Bedrock we now get the following error when using any of the default metrics defined under [`mlflow.metrics.genai.metric_definitions`](https://github.com/harupy/mlflow/blob/fix-Endpoint/mlflow/metrics/genai/metric_definitions.py):

> 'Failed to score model on payload. Error: 422 Client Error: Unprocessable Entity for url: http://127.0.0.1:5000/endpoints/completions/invocations. Response text: {"detail":"Cannot set both \'temperature\' and \'top_p\' parameters. Please use only the temperature parameter for your query."}'

I believe this is due to [`mlflow.metrics.genai.prompts.v1.default_parameters`](https://github.com/harupy/mlflow/blob/fix-Endpoint/mlflow/metrics/genai/prompts/v1.py#L10) being hardcoded as:

`default_parameters = {
    "temperature": 0.0,
    "max_tokens": 200,
    "top_p": 1.0,
}
`
This does not seem to be compatible with Claude models provided via AWS Bedrock.
Changing this to the following does work:

`default_parameters = {
    "temperature": 0.0,
    "max_tokens": 200,
}
`

however, there is currently no friendly way to modify these default parameters at runtime through the API when initialising a metric function under [`mlflow.metrics.genai.metric_definitions`](https://github.com/harupy/mlflow/blob/fix-Endpoint/mlflow/metrics/genai/metric_definitions.py).

Making this possible e.g. by adding a `parameters` arg as follows could fix this:

```
from mlflow.metrics.genai import relevance
relevance_metric = relevance(model="endpoints:/completions", parameters={"temperature": 0.0})
```

i've set up my `config.yaml` as follows:

```
endpoints:
  - name: completions
    endpoint_type: llm/v1/completions
    model:
      provider: bedrock
      name: anthropic.claude-v2
      config:
        aws_config: 
          aws_secret_access_key: <redacted>
          aws_access_key_id: <redacted
          aws_region: us-east-1
    limit:
      renewal_period: minute
      calls: 10

```
